File: app/src/main/java/com/mahesh/mvpsample/detail/DetailPresenter.java
package com.mahesh.mvpsample.detail;

import com.mahesh.mvpsample.detail.adapter.DetailAdapterModel;

public class DetailPresenter implements DetailContract.Presenter, DetailContract.AdapterPresenter {

    private DetailContract.View view;
    private DetailContract.ModelInteractor modelInteractor;

    DetailPresenter(DetailContract.View view) {
        this.view = view;
        this.modelInteractor = new DetailModelInteractor();
    }

    @Override
    public DetailAdapterModel getAdapterEntity(int position) {
        return modelInteractor.getAdapterEntityList().get(position);
    }

    @Override
    public int getAdapterEntityCount() {
        return modelInteractor.getAdapterEntityList().size();
    }

    @Override
    public void resume(DetailContract.View view) {
        this.view = view;
    }

    @Override
    public void pause() {
        view = null;
    }

    @Override
    public void create() {
        modelInteractor.createAdapterEntityList(new DetailContract.CreateDataCallback() {
            @Override
            public void onDataCreated() {
                if (isViewAvailable())
                    view.notifyAdapterEntityChanged();
            }
        });
    }

    @Override
    public Boolean isViewAvailable() {
        return view != null;
    }

    @Override
    public void adapterEntityClicked(int position) {
        if (isViewAvailable()) {
            view.showToast("Clicked " + getAdapterEntity(position).getText());
        }
    }
}

File: app/src/main/java/com/mahesh/mvpsample/detail/DetailPresenter.kt
package com.mahesh.mvpsample.detail

import com.mahesh.mvpsample.detail.DetailContract.AdapterPresenter
import com.mahesh.mvpsample.detail.adapter.DetailAdapterModel

class DetailPresenter(
    view: DetailContract.View
) : DetailContract.Presenter, AdapterPresenter {

    private var view: DetailContract.View? = view
    private val modelInteractor: DetailContract.ModelInteractor


    override val isViewAvailable: Boolean
        get() = view != null

    init {
        modelInteractor = DetailModelInteractor()
    }

    override fun getAdapterEntity(position: Int): DetailAdapterModel {
        return modelInteractor.adapterEntityList[position]
    }

    override val adapterEntityCount: Int
        get() = modelInteractor.adapterEntityList.size

    override fun resume(view: DetailContract.View?) {
        this.view = view
    }

    override fun pause() {
        view = null
    }

    override fun create() {
        modelInteractor.createAdapterEntityList {
            if (isViewAvailable){
                view?.notifyAdapterEntityChanged()
            }
        }
    }


    override fun adapterEntityClicked(position: Int) {
        if (isViewAvailable) {
            view!!.showToast("Clicked " + getAdapterEntity(position)?.text)
        }
    }
}
